### PR TITLE
BUG: Fix Series doesn't work in pd.astype(). Now treat Series as dict.

### LINKED ISSUE #16717: Using Series in DataFrame.astype() does not work correctly
#### Code Sample

```python
import pandas as pd
import numpy as np

df = pd.DataFrame([[1,2],[3,4]], columns=["A", "B"])


df

 	A 	B
0 	1 	2
1 	3 	4


df.astype({"A": np.float64, "B": np.float64})

 	A 	B
0 	1.0 	2.0
1 	3.0 	4.0

df.astype(pd.Series({"A": np.float64, "B": np.float64})) # using a series does not react at all (error or right behaviour)
 	A 	B
0 	1 	2
1 	3 	4

df.astype(list(pd.Series({"A": np.float64, "B": np.float64}))) # using a type that is not allowed raises error
[...] TypeError: data type not understood

```
#### Problem description
I wanted to parse the columns of a DF with a Series (instead of a dict as stated in the docs). I realised that the types did not change at all nor an error was raised. Casting to dict helped (see in code). 

#### Expected Output
I expect to give the same error as when a wrong data type (e.g. list) is used, or alternatively that it works the same way as a dict. At the moment nothing happens. 
 
<details>
pandas: 0.20.2
pytest: None
pip: 9.0.1
setuptools: 28.8.0
Cython: None
numpy: 1.13.0
scipy: None
xarray: None
IPython: 6.1.0
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by BranYang**
- [x] closes #16717 
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/frame/test_dtypes.py [hunk 1]
```diff
@@ -487,6 +487,54 @@ def test_astype_dict(self):
         assert_frame_equal(df, equiv)
         assert_frame_equal(df, original)
 
+    def test_astype_Series(self):
+        # GH16717
```
Comment (jreback) [inline pandas/tests/frame/test_dtypes.py]: I would instead modify the previous the test and use parametrize

File: doc/source/whatsnew/v0.21.0.txt [hunk 2]
```diff
@@ -91,6 +91,8 @@ Bug Fixes
 
 Conversion
 ^^^^^^^^^^
+- Fix a bug when pd.astype() receive Series as `dtype` paramter, no action is taken. Now Series
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: ``Bug in :func:`DataFarme.astype` where a passed Series as a dtype mapping would be ignored``

Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: Add the issue number as well

File: pandas/tests/frame/test_dtypes.py (hunk 1 above)
Comment (jreback) [inline pandas/tests/frame/test_dtypes.py]: then just add the issue number as an additional comment

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_dtypes.py [hunk 3]
```diff
@@ -494,6 +494,13 @@ def test_astype_dict_like(self, dtype_class):
         assert_frame_equal(df, equiv)
         assert_frame_equal(df, original)
 
+        # if dtypes provided is empty, the resulting DataFrame
+        # should be the same as the original DataFrame
+        dt7 = dtype_class({})
```
Comment (jreback) [inline pandas/tests/frame/test_dtypes.py]: add issue number as a comment
use result = 
rather than equiv

File: doc/source/whatsnew/v0.21.0.txt [hunk 4]
```diff
@@ -92,6 +92,8 @@ Bug Fixes
 
 Conversion
 ^^^^^^^^^^
+- Fix a bug that nothing happen when `Series` is passed to pd.astype() as `dtype`. Now `Series` 
+works the same as a `dict` when passed to `dtype`. (:issue:`16717`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: reword to be less verbose

**Review by jreback — APPROVED** [review]

File: pandas/core/generic.py [hunk 5]
```diff
@@ -3505,12 +3505,12 @@ def astype(self, dtype, copy=True, errors='raise', **kwargs):
...
         """
-        if isinstance(dtype, collections.Mapping):
+        if is_dict_like(dtype):
             if self.ndim == 1:  # i.e. Series
-                if len(dtype) > 1 or list(dtype.keys())[0] != self.name:
+                if len(dtype) > 1 or self.name not in dtype.keys():
                     raise KeyError('Only the Series name can be used for '
                                    'the key in Series dtype mappings.')
-                new_type = list(dtype.values())[0]
+                new_type = dtype.get(self.name)
```
Comment (jreback) [inline pandas/core/generic.py]: ``new_type = dtype[self.name]`` is ok here now

File: doc/source/whatsnew/v0.21.0.txt (hunk 4 above)
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: you can put this in 0.20.3

File: doc/source/whatsnew/v0.20.3.txt [hunk 6]
```diff
@@ -46,6 +46,7 @@ Conversion
 
 - Bug in pickle compat prior to the v0.20.x series, when ``UTC`` is a timezone in a Series/DataFrame/Index (:issue:`16608`)
 - Bug in Series construction when passing a Series with ``dtype='category'`` (:issue:`16524`).
+- Bug in pd.astype() when passing `Series` as `dtype`. (:issue:`16717`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.3.txt]: Bug in ``.astype()`` when passing a ``Series`` to ``dtype`` (use double backticks)

**Review by jreback — APPROVED** [review]
just a whatsnew fix. ping on green.

**Comment by BranYang** [comment]
@jreback Done as your suggestion.

**Comment by jreback** [comment]
thanks!

**Pull request merged**